### ISSUE TITLE
1-based indexing for fit.flatnames

### ISSUE THREAD

**Issue by ahartikainen** [stan-dev/pystan#430]
#### Summary:

Stan language uses 1-based indexing, so `fit.flatnames` should follow this. Currently summary and stansummary output vectors and matrices with 0-based indices.

#### Reproducible Steps:

`print(fit)` with matrices in `fit`

#### Expected Output:

1-based indexing.

**Comment by ariddell** [comment]
I'm torn here. I expect fit.extract()['theta'][0] to match `theta[0]`. Maybe we should poll users on this.

**Comment by ahartikainen** [comment]
I guess this is a "problem" between Stan lang and Python.

At the same time it would be great to use `fit.extract()['theta'][3]` --> `theta[3]`. On the other hand, I usually think in Stan lang, so if I change something in there, it would be the same.

This would also bring printing the fit closer to CmdStan/RStan.

The reason I created this issue is that the #425 is now implementing 1-based indexing. And if we decide to keep on using 0-based indexing, then we need to change the code for pandas (also same applies for the plotting code).

In the end, I think being consistent is the most important thing.

**Comment by ariddell** [comment]
I think in Stan lang until I have a numpy array, in which case I'm in the land of 0-offset indexing.

@betanalpha do you have any thoughts on this?

Perhaps we should post the before/after change on the discussion forum. This seems like a quasi-breaking change to me. It's definitely going to surprise everyone who got used to the old system.

**Comment by betanalpha** [comment]
I think it's naturally to use the native indexing for whatever data type is being used.  In particular, zero-indexing for NumPy arrays but one-indexing for the variable names of arrays, vectors, and matrices within the fit object.

**Comment by ariddell** [comment]
Sounds good. Shall we post a before version and an after version of the
summary on the discussion forum? We could also post an example of column
labels for the new pandas data frame.

I'm mainly worried that somebody is relying on the current behavior for
something important. I think we should give them warning in PyStan 2.18
and then change it in the next release.

**Comment by ahartikainen** [comment]
Should it contain also other changes (diagnostics + plotting).

**Comment by betanalpha** [comment]
We’ll never completely avoid surprised and/or upset users,
but the more channels on which changes are communicated
the better.

**Comment by jjramsey** [comment]
> In particular, there shouldn't be much conflict between having multiple conversions like this if containers are iterated over using list comprehensions and the indices aren't often called explicitly, right?

Be careful here. It's probably *not* a good idea to assume that indices aren't often called explicitly. In particular, if one is iterating over several arrays at once that are all of the same length, then one of the more obvious and common ways to do iteration is like this:

```
for i in range(foo):
    do_something(foo[i])
    do_something_else(bar[i])
```

Yes, there are often ways to avoid explicit indexing, but they can be less clear and more trouble than they're worth.

**Comment by ahartikainen** [comment]
Here is an example

    import pystan
    model_code ="""
    data {
      int<lower=0> J;
      real y[J];
      real<lower=0> sigma[J];
    }

    parameters {
      real mu;
      real<lower=0> tau;
      real theta[J];
    }

    model {
      mu ~ normal(0, 5);
      tau ~ cauchy(0, 5);
      theta ~ normal(mu, tau);
      y ~ normal(theta, sigma);
    }
    """

    stan_model = pystan.StanModel(model_code=model_code)
    schools_dat = {'J': 8,
               'y': [28,  8, -3,  7, -1,  1, 18, 12],
               'sigma': [15, 10, 16, 11,  9, 11, 10, 18]}
    fit = stan_model.sampling(data = schools_dat)

And then the output:

    # print(fit)
    
    Inference for Stan model: anon_model_68f43e2dc885e62bb7e4c7117e91aa73.
    4 chains, each with iter=2000; warmup=1000; thin=1; 
    post-warmup draws per chain=1000, total post-warmup draws=4000.

               mean se_mean     sd   2.5%    25%    50%    75%  97.5%  n_eff   Rhat
    mu         4.54    0.21   3.36   -2.2   2.34    4.5   6.78  10.71    247   1.02
    tau        3.88    0.16   2.98   0.79   1.73   3.05   5.18  11.73    340   1.01
    theta[0]   6.44    0.21   5.65  -3.43   2.96   5.94   9.61  19.79    729   1.01
    theta[1]   5.07    0.19   4.75  -4.57   2.14   5.07   8.21  14.42    646   1.01
    theta[2]   4.03    0.22   5.58   -7.5   1.01   4.23   7.43  14.08    620   1.01
    theta[3]   4.84     0.2   4.81  -4.93   1.85   4.81   7.93   13.9    593   1.01
    theta[4]   3.64     0.2    4.8  -6.61   0.91   3.95    6.8  12.12    596   1.01
    theta[5]   4.18     0.2   5.04  -6.23   1.25   4.26    7.5  13.14    659   1.01
    theta[6]    6.5     0.2   5.13  -2.67    3.3   6.14   9.25  18.68    640   1.01
    theta[7]   4.98    0.16   5.28  -5.12   1.84   4.91   8.24  16.21   1028   1.01
    lp__     -15.43    0.48    5.7 -26.53  -19.5 -15.45 -11.09   -4.8    142   1.03

    Samples were drawn using NUTS at Thu Feb 15 23:17:51 2018.
    For each parameter, n_eff is a crude measure of effective sample size,
    and Rhat is the potential scale reduction factor on split chains (at 
    convergence, Rhat=1).

And for updated fit

    # print(fit)
    
    Inference for Stan model: anon_model_68f43e2dc885e62bb7e4c7117e91aa73.
    4 chains, each with iter=2000; warmup=1000; thin=1; 
    post-warmup draws per chain=1000, total post-warmup draws=4000.

               mean se_mean     sd   2.5%    25%    50%    75%  97.5%  n_eff   Rhat
    mu         4.21    0.13   3.27  -2.37   2.01    4.2   6.53  10.55    624   1.01
    tau        3.96     0.2   3.27   0.57    1.6   3.06    5.3  12.54    269   1.01
    theta[1]   6.27    0.21   5.86  -3.22   2.57   5.76   9.02  20.14    760    1.0
    theta[2]   4.88    0.15   4.78  -4.34   1.88   4.83   7.57  14.95   1009   1.01
    theta[3]   3.66    0.17   5.44  -8.28   0.75   3.91   7.02  13.71    992   1.01
    theta[4]   4.57    0.14   4.81  -4.97   1.63   4.57   7.42  14.44   1133    1.0
    theta[5]   3.33    0.16    4.8  -7.03   0.52   3.55   6.53  12.22    899   1.01
    theta[6]   3.79    0.15   4.96  -6.99    0.8   4.03   6.95  13.35   1095   1.01
    theta[7]   6.42    0.21   5.23  -2.56   2.97   5.98   8.93  19.45    618   1.01
    theta[8]   4.72    0.15   5.31  -5.91    1.6   4.69    7.6  16.01   1325    1.0
    lp__     -15.18     0.5   6.31 -26.78 -19.54 -15.55  -10.7  -2.89    162   1.02

    Samples were drawn using NUTS at Thu Feb 15 23:22:20 2018.
    For each parameter, n_eff is a crude measure of effective sample size,
    and Rhat is the potential scale reduction factor on split chains (at 
    convergence, Rhat=1).

**Comment by ahartikainen** [comment]
@ariddell is it still fine if we use 1-based indexing in `.to_pandas`?

**Comment by ariddell** [comment]
I think we should be consistent. I'm fine with changing to 1-based
indexing but we should do it all at once.

This is easy to fix, right? We just need to change that flag in the
flatnames function?

**Comment by ahartikainen** [comment]
Yes, I still need to check that if the 0-based vs 1-based indexing is mentioned anywhere. If not then I guess we only need to comment that on the release notes? And maybe add a comment for the source code (for `flatnames` and `fnames_oi` sections, that we changed from the 0-based to 1-based indexing).

In my opinion, let's add it to 2.18, I'm not even sure, if anybody uses those names other than in visuals, (and if they are it's only `-1` to the code, but even then I'm not sure if it's a good coding habit to follow; I mean to parse the names or use `eval`).

**Comment by ahartikainen** [comment]
@bletham will this change have any effect on fbprophet?

**Comment by bletham** [comment]
@ahartikainen Thanks for checking. It does not seem to - we extract the parameters directly into numpy arrays with e.g. `fit['theta']` and don't index in the fit object.

**Comment by ariddell** [comment]
@ahartikainen I just tried out the to_dataframe function and noticed that it's already using the 1-based indexing in the column names. Is it getting its flatnames from somewhere else?

Here's the output:

```
             mean se_mean     sd   2.5%    25%    50%    75%  97.5%  n_eff   Rhat
alpha[0,0]  -0.01    0.01   0.99  -1.99  -0.69  -0.02   0.65   1.96   4552    1.0
alpha[1,0]  -0.02    0.01    1.0  -1.92  -0.69  -0.02   0.64   1.99   5671    1.0
alpha[0,1]   0.02    0.01   0.98  -1.93  -0.63   0.02   0.67    1.9   5540    1.0
alpha[1,1] 3.1e-3    0.01   0.98  -1.96  -0.64  -0.01   0.63   1.94   5438    1.0
alpha[0,2]  -0.03    0.01   1.01  -1.98   -0.7  -0.06   0.66    1.9   5889    1.0
alpha[1,2]  -0.01    0.01   1.02  -2.01  -0.69  -0.01   0.67   2.02   5732    1.0
beta[0]   -2.1e-3    0.02   1.38  -2.72  -0.93   0.01   0.93   2.69   5789    1.0
beta[1]      0.04    0.02   1.44  -2.71  -0.96   0.05   1.02   2.82   5920    1.0
lp__        -3.97    0.05   2.06  -9.14  -5.02  -3.62  -2.47  -1.01   1559    1.0
```

and the columns of the ``to_dataframe``:

```
Index(['alpha_1_1', 'alpha_2_1', 'alpha_1_2', 'alpha_2_2', 'alpha_1_3',
       'alpha_2_3', 'beta_1', 'beta_2', 'lp__'],
      dtype='object')
```

I'm mainly thinking that they should be getting their parameter name information from the same source.

**Comment by ahartikainen** [comment]
It is getting it inside the function.

The function parses data from the raw fit object and then reshapes that to table. It more natural to do the naming inplace than trying to get the name from `.flatnames` in this case and the do the renaming tricks. 

Also the naming `alpha_1_1`instead of `alpha[1,1]` is more suitable for pandas. You can call column names as a method.

e.g.

    `df.alpha_1_1`

**Comment by ahartikainen** [comment]
#431 fixed this.

Going back to `[1,1]` for pandas is handled in issue #443 and is going to be fixed with #444.

Closing this issue.
